ncclient/transport/third_party/junos/parser.py

```diff
@@ -0,0 +1,238 @@
+# Copyright 2018 Nitin Kumar
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+from threading import Lock
+
+from xml.sax.handler import ContentHandler
+from lxml import etree
+from lxml.builder import E
+from xml.sax._exceptions import SAXParseException
+from xml.sax import make_parser
+
+from ncclient.transport.parser import DefaultXMLParser
+from ncclient.operations import rpc
+
+import logging
+logger = logging.getLogger("ncclient.transport.third_party.junos.parser")
+
+# v1.0: RFC 4742
+MSG_DELIM = "]]>]]>"
+MSG_DELIM_LEN = len(MSG_DELIM)
+
+RPC_REPLY_END_TAG = "</rpc-reply>"
+RPC_REPLY_END_TAG_LEN = len(RPC_REPLY_END_TAG)
+
+
+class JunosXMLParser(DefaultXMLParser):
+    def __init__(self, session):
+        super(JunosXMLParser, self).__init__(session)
+        self._session = session
+        self.sax_parser = make_parser()
+        self.sax_parser.setContentHandler(SAXParser(session))
+
+    def parse(self, data):
+        try:
+            self.sax_parser.feed(data)
+        except SAXParseException:
+            self._delimiter_check(data)
+        finally:
+            self._parse10()
+
+    def _delimiter_check(self, data):
+        """
+        SAX parser throws SAXParseException exception, if there is extra data
+        after MSG_DELIM
+
+        :param data: content read by select loop
+        :return: None
+        """
+        data = data.decode('UTF-8')
+        if MSG_DELIM in data:
+            # need to feed extra data after MSG_DELIM
+            msg, delim, remaining = data.partition(MSG_DELIM)
+            self._session._buffer.seek(0, os.SEEK_END)
+            self._session._buffer.write(delim.encode())
+            # we need to renew parser, as old parser is gone.
+            self.sax_parser = make_parser()
+            self.sax_parser.setContentHandler(SAXParser(self._session))
+            if remaining.strip() != '':
+                self.sax_parser.feed(remaining)
+        elif RPC_REPLY_END_TAG in data:
+            msg, delim, remaining = data.partition(RPC_REPLY_END_TAG)
```

Comment:
Two more scenarios are not yet considered.

Scenario 1: When data is "-reply/>]]>" 

Data is not MSG_DELIM so first if will not match. Data is not <rpc-reply/> so second will not match. Third else will match and nothing is done and dispatch will never be called. 

Scenario 2: Data is '</rpc-reply>]]>'

Second if will match. ]]> will be written to _buffer. parse10() will be called. But since delimiter is not full arrived dispatch will not be called. In the next iteration remaining delimiter will come in data and it will not match both if and elif. It will match else. Nothing is done and dispatch will never get called.